src/stitching/delegateToSchema.ts

```diff
@@ -22,113 +22,148 @@ import FilterToSchema from '../transforms/FilterToSchema';
-    Object.keys(info.fragments).map(
-      fragmentName => info.fragments[fragmentName],
-    ),
-    info.operation.variableDefinitions,
+export function createBatchOperation(
+  targetSchema: GraphQLSchema,
+  targetOperation: 'query' | 'mutation' | 'subscription',
+  rootDefs: { [key: string]: [{ [key: string]: any }, { [key: string]: any }] },
+  graphqlContext: { [key: string]: any },
+  documentInfo: {
+    operation: {
+      name?: { [key: string]: any }
+      variableDefinitions?: Array<VariableDefinitionNode>,
+    },
+    variableValues?: { [variableName: string]: any },
+    fragments?: { [fragmentName: string]: FragmentDefinitionNode },
+  },
+  transforms?: Array<Transform>,
+): FetcherOperation {
+  const roots = Object.keys(rootDefs).map(key => {
+    const [args, info] = rootDefs[key];
+    const [a, n] = key.split(':');
+    const name = n || a;
+    const alias = n ? a : null;
+    return {
+      key: alias || name,
+      name,
+      alias,
+      args,
+      info: info || documentInfo
+    };
+  });
+
+  const selections: Array<SelectionNode> = roots.reduce((newSelections, { key, name: rootFieldName, info, alias, args }) => {
+    const rootSelections = info.fieldNodes.map((selection: FieldNode) => {
+       if (selection.kind === Kind.FIELD) {
+         const rootField: FieldNode = {
+           kind: Kind.FIELD,
+           name: {
+             kind: Kind.NAME,
+             value: rootFieldName,
+           },
+           alias: alias
+            ? {
+              kind: Kind.NAME,
+              value: alias
+            }
+            : null,
+           arguments: selection.arguments,
+           selectionSet: selection.selectionSet
+         };
+         return rootField;
+       }
+       return selection;
+    });
+
+    return newSelections.concat(rootSelections);
+  }, []);
+
+  const selectionSet: SelectionSetNode = {
+    kind: Kind.SELECTION_SET,
+    selections,
+  };
+
+  const operationDefinition: OperationDefinitionNode = {
+    kind: Kind.OPERATION_DEFINITION,
+    operation: targetOperation,
+    variableDefinitions: documentInfo.operation.variableDefinitions,
+    selectionSet,
+  };
+
+  const fragments = Object.keys(documentInfo.fragments).map(
+    fragmentName => documentInfo.fragments[fragmentName],
   );
 
+  const document = {
+    kind: Kind.DOCUMENT,
+    definitions: [operationDefinition, ...fragments],
+  };
+
   const rawRequest: Request = {
-    document: rawDocument,
-    variables: info.variableValues as Record<string, any>,
+    document,
+    variables: documentInfo.variableValues as Record<string, any>,
   };
 
-  const transforms = [
-    ...(options.transforms || []),
-    AddArgumentsAsVariables(options.schema, args),
-    FilterToSchema(options.schema),
-    AddTypenameToAbstract(options.schema),
-    CheckResultAndHandleErrors(info, options.fieldName),
+  transforms = [
+    ...(transforms || []),
+    ...roots.map(({ args }) => AddArgumentsAsVariables(targetSchema, args)),
```

Comment:
This will reset a variable counter each time. I think you should have one `AddArgumentsAsVariables`, it should be modified accept args lists with fieldName or alias and assign variables to them.

Reply:
This way we don't need to traverse AST that many times.

Reply:
@freiksenet do you think the type should be:
```ts
args: { [key: string]: any } | Array<{ fieldName: string, alias?: string, args: { [key: string]: any } }>
```
? 
That seems to be the only way we can support both patterns right?